### ISSUE TITLE
Thread group count is above the maximum allowed limit

### ISSUE THREAD

**Issue by SaschaDeWaal** [Unity-Technologies/barracuda-release#67]
I am trying out a pre-trained model (https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/duc)

But I keep getting an error that the thread group is above the maximum allowed limit.
Any tips or direction on how I could fix it?

```
Thread group count is above the maximum allowed limit. Maximum allowed thread group count is 65535.
UnityEngine.ComputeShader:Dispatch(Int32, Int32, Int32, Int32)
Barracuda.ComputeFunc:Dispatch(Int32, Int32, Int32) (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/BarracudaReferenceCompute.cs:1422)
Barracuda.ComputeFunc:Dispatch(Int32[]) (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/BarracudaReferenceCompute.cs:1408)
Barracuda.ComputeKernel:Dispatch() (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/BarracudaCompute.cs:435)
Barracuda.ComputeOps:Softmax(Tensor) (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/BarracudaCompute.cs:847)
Barracuda.StatsOps:Barracuda.IOps.Softmax(Tensor) (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/StatsOps.cs:209)
Barracuda.<ExecuteAsync>d__27:MoveNext() (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/GenericWorker.cs:587)
Barracuda.GenericWorker:Execute() (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/GenericWorker.cs:117)
Barracuda.GenericWorker:Execute(IDictionary`2) (at Library/PackageCache/com.unity.barracuda@0.4.0-preview/Barracuda/Core/Backends/GenericWorker.cs:105)
<Working>d__10:MoveNext() (at Assets/Scripts/ModelTester.cs:59)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

```

```
    using Barracuda;
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using UnityEngine;
     
    public class ModelTester : MonoBehaviour
    {
        private const string INPUT_NAME = "data";
     
        [SerializeField]
        private NNModel modelFile;
     
        [SerializeField]
        private Texture2D texture;
     
        private IWorker worker;
     
        private void Start() {
            Model model = ModelLoader.Load(this.modelFile, false);
     
            this.worker = WorkerFactory.CreateWorker(WorkerFactory.Type.ComputePrecompiled, model);
            StartCoroutine(Working());
        }
     
        private IEnumerator Working() {
            Debug.Log("Starting");
            yield return new WaitForSeconds(1);
     
            using (var tensor = new Tensor(texture)) {
                Dictionary<string, Tensor> inputs = new Dictionary<string, Tensor>();
     
                try {
                    inputs.Add(INPUT_NAME, tensor);
     
                    this.worker.Execute(inputs);
                } catch (Exception ex) {
                    Debug.LogError(ex);
                }
     
                // Do stuff with the result
                Debug.Log("Done");
            }
        }
    }
```

**Comment by AlexRibard** [comment]
How big is the input? 3x800x800 ? Can you try with a smaller input?
The issue is on us, we should unwrap our Softmax dispatch.
I'll let you know when this gets fixed

**Comment by SaschaDeWaal** [comment]
Thanks for the reply AlexRibard,

Yes, the input is 3x800x800.
I tried 256,  64, and 32 and I always get this error.

So might it be a problem with the model? I am not that experienced with this.

**Comment by AlexRibard** [comment]
No, no the problem is definitely us. I was trying to get a workaround until we fix the issue.
Let me explain the problem:
![image](https://user-images.githubusercontent.com/12990800/86032310-206f7680-ba05-11ea-9132-0706e45fef37.png)
So this is the end of the model
Internally for `Softmax` we dispatch a compute shader on H*W*C and that has a hardware limit of 65535. In your case that is equal to 19*160000 = 3040000 :)
For activations, for large input we unwrap the dispatches to make it work. But for softmax we didn't implement it.
Why I was trying to test with a lower input dim is that the output H*W*C might be < than 65535.

Can you try to remove the Softmax at the end and do it on the CPU?
If you only have the .onnx model, try something like that in python

```
model = onnx.load(modelName)

model.graph.output.clear();

output = helper.ValueInfoProto()
output.name = o
model.graph.output.append(output)

onnx.save(model, outputName)
```

If this is too convoluted, we should have a fix by the end of the week (hopefully)

**Comment by SaschaDeWaal** [comment]
Because this is still in the preview state (and currently not working), I started to use Microsoft's ML.Net instead.

But thanks a lot for your help!
